File: master/strstr.py

class Solution:
    # @param haystack, a string
    # @param needle, a string
    # @return a string or None
    def strStr(self, haystack, needle):
        begin_h = 0
        curr_h = 0
        curr_n = 0
        while curr_h<len(haystack):
            if curr_n == len(needle):
                return haystack[begin_h:]
            if haystack[curr_h] == needle[curr_n]:
                curr_h += 1
                curr_n += 1
            else:
                begin_h += 1
                curr_h = begin_h
                curr_n = 0
        return None if curr_n<len(needle) else haystack[begin_h:]

print Solution().strStr("hello, world", "o, world")
print Solution().strStr("hello, world", "o, word")
print Solution().strStr("NULL", "NOT NULL")

File: master/strstr.cpp
#include "header.h"

char *strStr(char *haystack, char *needle) {
    if(!needle) return NULL;
    if(!*needle) return haystack;
    char* h_head = haystack;
    char* h_curr = h_head;
    char* n_curr=needle;
    while(*h_curr){
        if(!*n_curr) return h_head;
        if(*n_curr != *h_curr){
            h_head++;
            h_curr = h_head;
            n_curr = needle;
        }
        else{
            n_curr++;
            h_curr++;
        }
    }
    return *n_curr?NULL:h_head;
}

int main(){
    printf("%s\n", strStr("hello, world", "o, world"));
    char* res = strStr("hello, world", "o, word");
    printf("%s\n", res?"NULL":"NOT NULL");
//    printf("%s\n", strStr("hello, world", ""));
    return 0;
}
